test-suite/blackformula.cpp

```diff
@@ -276,6 +277,182 @@ void BlackFormulaTest::testImpliedVolAdaptiveSuccessiveOverRelaxation() {
+                        << "\n analytical delta  :" << delta
+                        << "\n approximated delta:" << deltaApprox);
+        }
+    }
+}
+
+void BlackFormulaTest::testBlackFormulaForwardDerivative() {
+
+    BOOST_TEST_MESSAGE("Testing forward derivative of the Black formula...");
+
+    std::vector<Real> strikes;
+    strikes.push_back(0.1);
+    strikes.push_back(0.5);
+    strikes.push_back(1.0);
+    strikes.push_back(2.0);
+    strikes.push_back(3.0);
+    const Real vol = 0.1;
+    assertBlackFormulaForwardDerivative(Option::Call, strikes, vol);
+    assertBlackFormulaForwardDerivative(Option::Put, strikes, vol);
+}
+
+void BlackFormulaTest::testBlackFormulaForwardDerivativeWithZeroStrike() {
+
+    BOOST_TEST_MESSAGE("Testing forward derivative of the Black formula "
+        "with zero strike...");
+
+    std::vector<Real> strikes;
+    strikes.push_back(0.0);
+    const Real vol = 0.1;
+    assertBlackFormulaForwardDerivative(Option::Call, strikes, vol);
+    assertBlackFormulaForwardDerivative(Option::Put, strikes, vol);
+}
+
+void BlackFormulaTest::testBlackFormulaForwardDerivativeWithZeroVolatility() {
+
+    BOOST_TEST_MESSAGE("Testing forward derivative of the Black formula "
+        "with zero volatility...");
+
+    std::vector<Real> strikes;
+    strikes.push_back(0.1);
+    strikes.push_back(0.5);
+    strikes.push_back(1.0);
+    strikes.push_back(2.0);
+    strikes.push_back(3.0);
+    const Real vol = 0.0;
+    assertBlackFormulaForwardDerivative(Option::Call, strikes, vol);
+    assertBlackFormulaForwardDerivative(Option::Put, strikes, vol);
+}
+
+void assertBachelierBlackFormulaForwardDerivative(
+    Option::Type optionType,
+    const std::vector<Real> &strikes,
+    Real bpvol)
+{
+    Real forward = 1.0;
+    Real tte = 10.0;
+    Real stdDev = bpvol * std::sqrt(tte);
+    Real discount = 0.95;
+    Real bump = 0.0001;
+    Real epsilon = 1.e-10;
+
+    for (std::vector<Real>::const_iterator it = strikes.begin(); it != strikes.end(); ++it)
+    {
+        Real strike = *it;
+        Real delta = bachelierBlackFormulaForwardDerivative(
+            optionType, strike, forward, stdDev, discount);
+        Real bumpedDelta = bachelierBlackFormulaForwardDerivative(
+            optionType, strike, forward + bump, stdDev, discount);
+
+        Real basePremium = bachelierBlackFormula(
+            optionType, strike, forward, stdDev, discount);
+        Real bumpedPremium = bachelierBlackFormula(
+            optionType, strike, forward + bump, stdDev, discount);
+        Real deltaApprox = bumpedPremium - basePremium;
+
+        /*! Based on the Mean Value Theorem, the below inequality
+            should hold for any function that is monotonic in the
+            area of the bump.
+         */
+        bool success = (std::max(delta * bump, bumpedDelta * bump) + epsilon > (
+            deltaApprox) > std::min(delta * bump, bumpedDelta * bump) - epsilon);
```

Comment:
```suggestion
        bool success =
            (std::max(delta * bump, bumpedDelta * bump) + epsilon > deltaApprox)
            && (deltaApprox > std::min(delta * bump, bumpedDelta * bump) - epsilon);
```

Reply:
Thank you for help! I changed it accordingly.